### ISSUE TITLE
Optimize context caching in the TestContext framework

### ISSUE THREAD

**Issue by seregamorph** [spring-projects/spring-framework#32289]
The Spring TestContext Framework (TCF) uses a pretty convenient and flexible approach to create and subsequently reuse the context by aggregate `MergedContextConfiguration`. However it has a drawback: in a large test suite there can eventually be created too many parallel active contexts, consuming a lot of resources like large thread pools or Testcontainers beans.

<img width="601" alt="Screenshot 2024-02-18 at 12 46 45" src="https://github.com/spring-projects/spring-framework/assets/2844909/a07cd4a4-7e08-4d13-b6b5-d7f544e91ff2">

There are several good practices to reduce the number of independent configurations like introducing common test super classes and reducing usage of `@MockBean` annotations. Also we can reduce the overhead of each new context like statically defined/reusable Testcontainers containers.

Unfortunately, these approaches do not work very well for big distributed projects with many teams contributing independently. So eventually OOM and other problems arise.

As a mitigation there can be some urgent fixes like using `@DirtiesContext` or the `spring.test.context.cache.maxSize=1` option as suggested by @snicoll (https://github.com/spring-projects/spring-boot/issues/15654). The suggested approach fixed the problem, but it has a disadvantage as well: the total test execution time increased, due to the larger number of context re-initializations.
  
I had the same problem while working with a https://miro.com monolith server application, and I've found two more approaches to reduce the number of active contexts.

## Smart (Auto) DirtiesContext

For single-threaded test executions, we can know the sequence (list) of tests in the very beginning of the suite. It's easy to calculate the `MergedContextConfiguration` per each class - and now it's possible to define a custom [SmartDirtiesTestExecutionListener](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/java/com/github/seregamorph/testsmartcontext/SmartDirtiesContextTestExecutionListener.java) with an `afterTestClass` implementation pretty similar to standard `DirtiesContextTestExecutionListener`, but there is binary logic: if the current test class is the last test class using a given context configuration, close the context by marking it as dirty.

This trivial approach significantly reduced the number of active contexts and decreased the time of test execution (as fewer resources like CPU were consumed).

The only problem was that on the level of the TCF it's not possible to access the suite, so I originally implemented a custom TestNG listener, and later the JUnit 5 implementation was added.
  
## Test reordering

We can do even better if the test execution sequence is reordered - so we can group test classes that share the same context configuration sequentially, and the number of active spring contexts will never exceed one.

The following chart demonstrates the approach (same color = same `MergedContextConfiguration`):

<img width="600" alt="reorder-and-smart-dirties-context" src="https://github.com/spring-projects/spring-framework/assets/2844909/dc8c6e24-79a5-4858-a17c-3c18cb2e31b6">

It's not possible to reorder tests on the level of the TCF, but it's possible to do so on the level of:

* JUnit Jupiter [SmartDirtiesClassOrderer](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/java/com/github/seregamorph/testsmartcontext/jupiter/SmartDirtiesClassOrderer.java) defined via [junit-platform.properties](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/resources/junit-platform.properties)
* TestNG [SmartDirtiesSuiteListener](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/java/com/github/seregamorph/testsmartcontext/testng/SmartDirtiesSuiteListener.java) defined via [META-INF/services ITestNGListener](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/resources/META-INF/services/org.testng.ITestNGListener)
* JUnit 4 vintage-engine [SmartDirtiesPostDiscoveryFilter](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/java/com/github/seregamorph/testsmartcontext/SmartDirtiesPostDiscoveryFilter.java) defined via [META-INF/services PostDiscoveryFilter](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/resources/META-INF/services/org.junit.platform.launcher.PostDiscoveryFilter) (workaround approach)

## Metrics: fewer contexts and faster

Here is a sample test suite:

<img width="598" alt="Screenshot 2024-02-18 at 12 53 24" src="https://github.com/spring-projects/spring-framework/assets/2844909/95e19f69-c1a8-47c2-88c5-686d6fdbe484">

On the horizontal axis there is a timeline and on the vertical axis the number of active spring contexts (calculated each 10 sec). As you can see, the Smart DirtiesContext + test reordering (yellow) is always better - it has fewer active contexts, and the total time of test execution is **the smallest** (because of less CPU consumption + minimal context re-initialization).

The following chart is about number of parallel active Testcontainers docker containers (represented as spring Beans) for another test suite and is even more representative (unfortunately I cannot compare with `cache.maxSize=1` approach): 
<img width="597" alt="Screenshot 2024-02-18 at 16 32 16" src="https://github.com/spring-projects/spring-framework/assets/2844909/6b81319c-886a-4010-853e-7241ae3c1ed3">

## Prototype

I've made a library https://github.com/seregamorph/spring-test-smart-context that implements this approach for JUnit Jupiter, TestNG, and even JUnit 4 via the vintage-engine to demonstrate the PoC.

@snicoll and @marcphilipp were so kind to give some initial feedback, and then Stéphane suggested to submit a ticket to continue discussion here.

I understand that the current implementation of the TCF conceptually does not allow this solution as it works on another level, but this can be a possible direction of library evolution (for both spring-framework and junit-platform). As this approach has significant advantages like flexibility and freedom for engineers - they do not need to care too much regarding the optimizations.

Spring team, curious about your opinion. cc @sbrannen

**Comment by sbrannen** [comment]
Hi @seregamorph,

Thanks for sharing your ideas as well as your prototype! 👍 

Those are indeed very interesting approaches to the task at hand.

I especially like the idea of executing test classes that share the same context configuration sequentially.

Though, we might be able to achieve a similar effect by tracking all test classes that use the same context configuration and closing after the last one, regardless of the order in which they are executed (for example, by decrementing a counter and eagerly closing the context once we hit 0).

I do have some concerns, however, and I'll add a few of them here as "food for thought".

- This would have to be an opt-in feature, since we could never apply these semantics by default. However, making it opt-in may be problematic since some of the techniques used in the prototype require that "services" be automatically registered based on what's in the classpath.
  - So, this may potentially require an additional artifact (JAR) that users can optionally add to the classpath, and we've never done that before in the core Spring Framework.
- I'm a bit hesitant to implement this feature set differently for JUnit Jupiter, JUnit 4, and TestNG.
  - In light of that, we may choose to only support this with JUnit Jupiter.
- Some parts of this won't work if...
  - a different `ClassOrderer` is used
  - test classes are executed in parallel
    - either by JUnit Platform / JUnit Jupiter
    - or by the IDE or build tool

In any case, I have assigned this to the 6.2.x milestone since I think it's worth investigating what's possible.

**Comment by snicoll** [comment]
> we might be able to achieve a similar effect by tracking all test classes that use the same context configuration and closing after the last one, regardless of the order in which they are executed

This has the downside of keeping the context (and all the resources attached to it) for a longer period of time. I think it's important to take into account what @seregamorph and team have been doing here. Speaking of which, I'd like to see how we can them to submit a PR based on our guidance given they already have an implementation.

**Comment by seregamorph** [comment]
> making it opt-in may be problematic since some of the techniques used in the prototype require that "services" be automatically registered based on what's in the classpath. So, this may potentially require an additional artifact (JAR) that users can optionally add to the classpath

> Some parts of this won't work if... a different ClassOrderer is used

That's right! Maybe for this reason it makes sense to plan possible changes in junit-platform first? Such changes can unify the approach for all three test frameworks (JUnit 4 via vintage-engine, Jupiter, TestNG via testng-engine; maybe spock).

**Comment by seregamorph** [comment]
One more possible approach that can allow to avoid introducing new jar artifacts in the group of spring-framework:
* keep current behaviour as is, for the sake of back compatibility
* add [SmartDirtiesClassOrderer](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/java/com/github/seregamorph/testsmartcontext/jupiter/SmartDirtiesClassOrderer.java) to spring-test TCF framework (not activated by default)
* add [SmartDirtiesContextTestExecutionListener](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/java/com/github/seregamorph/testsmartcontext/SmartDirtiesContextTestExecutionListener.java) to the list of default `TestExecutionListeners` and make it only working if `SmartDirtiesClassOrderer` was executed
* document new behaviour for projects that have issues with too many active contexts, which will require just to define [resources/junit-platform.properties](https://github.com/seregamorph/spring-test-smart-context/blob/master/spring-test-smart-context/src/main/resources/junit-platform.properties) with content:
```
junit.jupiter.testclass.order.default = com.github.seregamorph.testsmartcontext.jupiter.SmartDirtiesClassOrderer
```

**Comment by seregamorph** [comment]
@sbrannen @marcphilipp thanks for your livestream presentation today.
I was asking about this issue. Actually I used a chance to ask you both :) as spring TCF cannot sort tests, so it should have some support on the side of junit-platform. So are there any plans like this on the side of the junit?

Also I see this issue was added to 6.2.x milestone, but it seems that 6.2.0 still does not have it.

**Comment by wimdeblauwe** [comment]
+1 for tackling this. I have had multiple projects where I need to set `spring.test.context.cache.maxSize=1` to avoid that my build fails on GitHub actions. Locally it is usually not a problem, but in the build pipeline it is (I assume due to lack of memory or cpu or both). Another alternative is using [custom ApplicationContextInitializer](https://maciejwalkowiak.com/blog/testcontainers-spring-boot-setup/#custom-application-context-initializer), then the same docker/testcontainers instance is shared across all contexts in the cache. But the `@Bean` model of using Testcontainers is so much nicer, so I would hate to have to go back to that "old way".

**Comment by seregamorph** [comment]
@wimdeblauwe have you tried https://github.com/seregamorph/spring-test-smart-context mentioned as a working prototype solving this problem? BTW it supports Spring Boot 2, 3, and 4, as well as both TestNG and JUnit 5 and 6.

**Comment by wimdeblauwe** [comment]
> [@wimdeblauwe](https://github.com/wimdeblauwe) have you tried https://github.com/seregamorph/spring-test-smart-context mentioned as a working prototype solving this problem? BTW it supports Spring Boot 2, 3, and 4, as well as both TestNG and JUnit 5 and 6.

I have not, was unaware that there was already an actual implementation to try. Thanks for pointing that out, I will surely try it!

**Comment by wimdeblauwe** [comment]
@seregamorph I just tested this now and it works great! Tests failed in CI before using this. Just adding the jar to the classpath made them work without issue now.

I hope the Spring team can figure it out so we get support for this natively, but until then, this is getting added to my default toolbox.

**Comment by seregamorph** [comment]
Glad to hear that. I continue working on this solution and there are many things planned (e.g. add resource leakage detector during the test execution, like increasing number of threads, heap memory consumption, etc.). It would be also nice to have a 🌟, and follow the updates ❤️.

Regarding the native support of this on the level of the spring framework the main issue is that Spring cannot control (and even be aware of) the actual order of the test execution, test engines like JUnit are not aware about Spring Context caching as a conception (they operate `Extension` high-level abstractions). Also there are several corner cases like parallel test execution, mixing several test engines in the same source set and few others. But the good news is that both JUnit Platform (most widely used in the Java ecosystem) and Spring Test are led by the same people who closely collaborate AFAIK, so 🤞 this can be solved somehow.
